File: biome_sim/noise/perlin2d.py
from __future__ import annotations

from dataclasses import dataclass

from biome_sim.core.prng import Mulberry32, clamp_u32


def _fade(t: float) -> float:
    return t * t * t * (t * (t * 6.0 - 15.0) + 10.0)


def _lerp(a: float, b: float, t: float) -> float:
    return a + t * (b - a)


def _grad(h: int, x: float, y: float) -> float:
    # 8 gradient directions based on lower bits.
    hh = h & 7
    u = x if hh < 4 else y
    v = y if hh < 4 else x
    return (u if (hh & 1) == 0 else -u) + (v if (hh & 2) == 0 else -v)


@dataclass(frozen=True)
class FbmOptions:
    octaves: int = 5
    lacunarity: float = 2.0
    persistence: float = 0.5


class Perlin2D:
    def __init__(self, seed: int):
        rng = Mulberry32(clamp_u32(seed))
        p = list(range(256))
        # Fisher-Yates shuffle
        for i in range(255, 0, -1):
            j = int(rng.random() * (i + 1))
            p[i], p[j] = p[j], p[i]
        self._perm = [p[i & 255] for i in range(512)]

    def noise(self, x: float, y: float) -> float:
        xi = int(x) & 255
        yi = int(y) & 255
        xf = x - int(x)
        yf = y - int(y)

        u = _fade(xf)
        v = _fade(yf)

        aa = self._perm[xi + self._perm[yi]]
        ab = self._perm[xi + self._perm[yi + 1]]
        ba = self._perm[xi + 1 + self._perm[yi]]
        bb = self._perm[xi + 1 + self._perm[yi + 1]]

        x1 = _lerp(_grad(aa, xf, yf), _grad(ba, xf - 1.0, yf), u)
        x2 = _lerp(_grad(ab, xf, yf - 1.0), _grad(bb, xf - 1.0, yf - 1.0), u)

        # Roughly in [-1, 1]; normalize by 1/sqrt(2) to keep gradients sane.
        return _lerp(x1, x2, v) * 0.7071067811865475

    def fbm(self, x: float, y: float, opts: FbmOptions) -> float:
        octaves = max(1, int(opts.octaves))
        freq = 1.0
        amp = 1.0
        s = 0.0
        amp_sum = 0.0

        for _ in range(octaves):
            s += self.noise(x * freq, y * freq) * amp
            amp_sum += amp
            freq *= float(opts.lacunarity)
            amp *= float(opts.persistence)

        if amp_sum == 0.0:
            return 0.0
        return s / amp_sum

File: src/terrain/Perlin2D.ts
import { mulberry32 } from '../core/SeededRandom'

function fade(t: number): number {
  return t * t * t * (t * (t * 6 - 15) + 10)
}

function lerp(a: number, b: number, t: number): number {
  return a + t * (b - a)
}

function grad(hash: number, x: number, y: number): number {
  // 8 gradient directions based on lower bits.
  const h = hash & 7
  const u = h < 4 ? x : y
  const v = h < 4 ? y : x
  return ((h & 1) === 0 ? u : -u) + ((h & 2) === 0 ? v : -v)
}

export type FbmOptions = {
  octaves: number
  lacunarity: number
  persistence: number
}

export class Perlin2D {
  private perm: Uint8Array

  constructor(seed: number) {
    const rng = mulberry32(seed)
    const p = new Uint8Array(256)
    for (let i = 0; i < 256; i++) p[i] = i
    for (let i = 255; i > 0; i--) {
      const j = Math.floor(rng() * (i + 1))
      const tmp = p[i]
      p[i] = p[j]
      p[j] = tmp
    }
    this.perm = new Uint8Array(512)
    for (let i = 0; i < 512; i++) this.perm[i] = p[i & 255]
  }

  noise(x: number, y: number): number {
    const xi = Math.floor(x) & 255
    const yi = Math.floor(y) & 255
    const xf = x - Math.floor(x)
    const yf = y - Math.floor(y)

    const u = fade(xf)
    const v = fade(yf)

    const aa = this.perm[xi + this.perm[yi]]
    const ab = this.perm[xi + this.perm[yi + 1]]
    const ba = this.perm[xi + 1 + this.perm[yi]]
    const bb = this.perm[xi + 1 + this.perm[yi + 1]]

    const x1 = lerp(grad(aa, xf, yf), grad(ba, xf - 1, yf), u)
    const x2 = lerp(grad(ab, xf, yf - 1), grad(bb, xf - 1, yf - 1), u)

    // Roughly in [-1, 1]
    return lerp(x1, x2, v) * 0.7071067811865475 // 1/sqrt(2) for gradient normalization
  }

  fbm(x: number, y: number, opts: FbmOptions): number {
    const octaves = Math.max(1, Math.floor(opts.octaves))
    let frequency = 1
    let amplitude = 1
    let sum = 0
    let ampSum = 0

    for (let i = 0; i < octaves; i++) {
      sum += this.noise(x * frequency, y * frequency) * amplitude
      ampSum += amplitude
      frequency *= opts.lacunarity
      amplitude *= opts.persistence
    }

    return ampSum === 0 ? 0 : sum / ampSum
  }
}
